### PR TITLE
register and login api

### PR DESCRIPTION
**Pull request by ahsanalam314**

### CODE REVIEW

File: controllers/user.controller.ts [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+        } catch (error) {
+            console.error(`UserController registerUser error: ${error}`);
+            throw new Error();
+        }
+    }
+
+    public async login(request: Request, response: Response) {
+        try {
+
+            const user = await UserService.login(request.body);
```
Comment (fahadsubzwari924) [inline controllers/user.controller.ts:28]: There should be a validation to validate the incoming request payload

File: services/user.service.ts [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+import { UserModel } from "../models";
+import { error, success } from "../helper/response.helper";
+import { ResponseMessage } from "../contants/response-message.contant";
+
+export class UserService {
+
+    public static async registerUser(user: any): Promise<any | undefined> {
+        try {
+
+            const newUser = new UserModel(user);
```
Comment (fahadsubzwari924) [inline services/user.service.ts]: The name shouldn't have the `Model` keyword. It should be just `User`.

File: services/user.service.ts [hunk 3]
```diff
@@ -0,0 +1,51 @@
+import { Request, Response } from "express";
+import { UserModel } from "../models";
+import { error, success } from "../helper/response.helper";
+import { ResponseMessage } from "../contants/response-message.contant";
+
+export class UserService {
+
+    public static async registerUser(user: any): Promise<any | undefined> {
```
Comment (fahadsubzwari924) [inline services/user.service.ts]: here, type shouldn't be any, there must be any rigid type here. Here type should be there for param `user` and also Promise type also should be there. And, this is not the right approach to have type like any | undefined. Type should be singular and not with undefined.

File: services/user.service.ts [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+            const {email, password} = user;
+
+            const userFound = await UserModel.findOne({ email });
+
+            if (!userFound) {
+                return error(ResponseMessage.User.emailNotFound);
+            }
+
+            const isMatch = await userFound.comparePassword(password);
```
Comment (fahadsubzwari924) [inline services/user.service.ts]: `comparePassword` shouldn't be the model level method. It should be in a service or util/helper.

File: services/user.service.ts [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+            return newUser;
+
+        } catch (error) {
+            console.error(`UserService registerUser error: ${error}`);
+            throw new Error();
+        }
+    }
+
+    public static async login(user: any): Promise<any | undefined> {
```
Comment (fahadsubzwari924) [inline services/user.service.ts]: here, type shouldn't be any, there must be any rigid type here. Here type should be there for param `user` and also Promise type also should be there. 
You can simply place the promise return type like this Promise<UserModel | undefined>

File: services/user.service.ts (hunk 2 above)
Comment (ahsanalam314) [reply services/user.service.ts]: done

File: services/user.service.ts [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+            return await newUser.save();
+        } catch (error) {
+            console.error(`UserService registerUser error: ${error}`);
+            throw new Error();
+        }
+    }
+
+    public static async findUserByEmail(email: string): Promise<IUserModel> {
+        try {
+            const user = await User.findOne({ email }) as IUserModel;
```
Comment (fahadsubzwari924) [inline services/user.service.ts]: This is useless assignment in user variable. You can just directly return from here

File: models/user.model.ts [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+        type: String,
+        required: true,
+    },
+    email: {
+        type: String,
+        required: true,
+        unique: true
+    },
+    password: {
+        type: String,
```
Comment (fahadsubzwari924) [inline models/user.model.ts:20]: there should be min & max length limit for password

File: models/user.model.ts [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+        required: true,
+    },
+    createdAt: {
+        type: Date,
+        required: true,
+        default: Date.now
+    },
+    updatedAt: {
+        type: Date,
```
Comment (fahadsubzwari924) [inline models/user.model.ts:31]: this shouldn't be `required`. There could be scenarios where you only have createdAt not updatedAt.

File: models/interface/user.model.interface.ts [hunk 9]
```diff
@@ -0,0 +1,11 @@
+import { Document } from 'mongoose';
+
+export interface IUserModel extends Document {
+    firstName: string;
+    lastName: string;
+    email: string;
+    password: string;
+    createdAt: Date;
+    updatedAt: Date;
+    comparePassword: (password: string) => Promise<boolean>;
```
Comment (fahadsubzwari924) [inline models/interface/user.model.interface.ts]: You can omit this comparePassword from the model. You can easily use Mongoose entity when using the `comparePassword` method. You don't need to include this in the interface. Interfaces are there to only reference or point to the correct return type.

File: services/user.service.ts (hunk 4 above)
Comment (ahsanalam314) [reply services/user.service.ts]: its done

File: services/user.service.ts (hunk 3 above)
Comment (ahsanalam314) [reply services/user.service.ts]: its done

File: models/user.model.ts (hunk 8 above)
Comment (ahsanalam314) [reply models/user.model.ts:31]: its is useful when user update its profile, name or password...

Comment (fahadsubzwari924) [reply models/user.model.ts:31]: I wanted to say that keep this field but as optional, not required one :)

File: controllers/user.controller.ts [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+                return response.status(404).json(errorResponse(ResponseMessage.User.emailNotFound));
+            }
+
+            const isMatch = await comparePassword(password, user.password);
+
+            if (!isMatch) {
+                return response.status(404).json(errorResponse(ResponseMessage.User.incorrectPassword));
+            }
+
+            return response.status(201).json(successResponse(ResponseMessage.User.loginSuccessfull, user));
```
Comment (fahadsubzwari924) [inline controllers/user.controller.ts:38]: HTTP status is wrong here, there should be a `200` HTTP status. `201` status if for returning in scenarios when new resource is being created successfully. Please update in the next PR.

**Review by fahadsubzwari924 — APPROVED** [review]

**Pull request merged**